xmtp_mls/src/storage/encrypted_store/group_intent.rs

```diff
@@ -1,25 +1,551 @@
+    deserialize::{self, FromSql, FromSqlRow},
+    expression::AsExpression,
+    serialize::{self, IsNull, Output, ToSql},
+    sql_types::Integer,
+    sqlite::Sqlite,
+};
+
+pub type ID = i32;
+
+#[repr(i32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, AsExpression, FromSqlRow)]
+#[diesel(sql_type = Integer)]
+/// Status of membership in a group, once a user sends a request to join
+pub enum IntentKind {
+    SendMessage = 1,
+    AddMembers = 2,
+    RemoveMembers = 3,
+    KeyUpdate = 4,
+}
 
-use super::schema::group_intents;
+#[repr(i32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, AsExpression, FromSqlRow)]
+#[diesel(sql_type = Integer)]
+/// Status of membership in a group, once a user sends a request to join
+pub enum IntentState {
+    ToPublish = 1,
+    Published = 2,
+    Committed = 3,
+}
 
-#[derive(Queryable, Identifiable, Debug, Clone)]
+#[derive(Queryable, Identifiable, Debug, PartialEq, Clone)]
 #[diesel(table_name = group_intents)]
 #[diesel(primary_key(id))]
 pub struct StoredGroupIntent {
-    pub id: i32,
-    pub kind: i32,
-    pub group_id: Vec<u8>,
+    pub id: ID,
+    pub kind: IntentKind,
+    pub group_id: group::ID,
     pub data: Vec<u8>,
-    pub state: i32,
+    pub state: IntentState,
     pub payload_hash: Option<Vec<u8>>,
     pub post_commit_data: Option<Vec<u8>>,
 }
 
-#[derive(Insertable, Debug, Clone)]
+impl_fetch!(StoredGroupIntent, group_intents, ID);
+
+#[derive(Insertable, Debug, PartialEq, Clone)]
 #[diesel(table_name = group_intents)]
 pub struct NewGroupIntent {
-    pub kind: i32,
+    pub kind: IntentKind,
     pub group_id: Vec<u8>,
     pub data: Vec<u8>,
-    pub state: i32,
+    pub state: IntentState,
+}
+
+impl_store!(NewGroupIntent, group_intents);
+
+impl NewGroupIntent {
+    pub fn new(kind: IntentKind, group_id: Vec<u8>, data: Vec<u8>) -> Self {
+        Self {
+            kind,
+            group_id,
+            data,
+            state: IntentState::ToPublish,
+        }
+    }
+}
+
+impl EncryptedMessageStore {
+    // Query for group_intents by group_id, optionally filtering by state and kind
+    pub fn find_group_intents(
+        &self,
+        conn: &mut DbConnection,
+        group_id: Vec<u8>,
+        allowed_states: Option<Vec<IntentState>>,
+        allowed_kinds: Option<Vec<IntentKind>>,
+    ) -> Result<Vec<StoredGroupIntent>, StorageError> {
+        let mut query = dsl::group_intents
+            .into_boxed()
+            .filter(dsl::group_id.eq(group_id));
+
+        if let Some(allowed_states) = allowed_states {
+            query = query.filter(dsl::state.eq_any(allowed_states));
+        }
+
+        if let Some(allowed_kinds) = allowed_kinds {
+            query = query.filter(dsl::kind.eq_any(allowed_kinds));
+        }
+
+        Ok(query.load::<StoredGroupIntent>(conn)?)
+    }
+
+    // Set the intent with the given ID to `Published` and set the payload hash. Optionally add `post_commit_data`
+    pub fn set_group_intent_published(
```

Comment:
Why not just mutate the struct and call `save()` on it, rather than having these three methods below? Not convinced the validation is worth the increased code surface/lower flexibility for change, but it's up to you

Reply:
We are going to have to enforce the state transitions somewhere. If we enforced elsewhere and saved, we would have to re-fetch the row from the db first since the local state might be out of date (or always use in a complicated tx that includes validation). This does everything in one SQL statement, which feels tidy. Can be used in a tx with the read or separately.